e2e/visual-tests/setup-container.js

```diff
@@ -0,0 +1,163 @@
+const Docker = require("dockerode");
+const path = require("path");
+
+const { runCommand } = require("./utils");
+
+const cwd = path.resolve(__dirname, "../../");
+const docker = new Docker();
+let container;
+
+const IMAGE_BASE_NAME = "metabase/cypress-runner";
+const IMAGE_TAG = "latest";
+const IMAGE_NAME = `${IMAGE_BASE_NAME}:${IMAGE_TAG}`;
+
+const BUILD_ARGS = [
+  `BINARY_ARCHITECTURE=${process.platform === "darwin" ? "arm64" : "x64"}`,
+]
+  .map(arg => ["--build-arg", arg])
+  .flat();
+
+const WORKING_DIRECTORY = "/app";
+const CONTAINER_BOUND_DIRECTORIES = ["/"];
+
+const SUCCESS_EXIT_CODE = 0;
+const ERROR_EXIT_CODE = 1;
+
+module.exports.setupContainer = async function setupContainer(options) {
+  const cleanup = prepareCleanup();
+
+  const exists = await isDockerImageCreated(IMAGE_NAME);
+
+  if (!exists) {
+    await buildImage();
+  }
+
+  const exitCode = await runContainer(options);
+
+  cleanup(exitCode);
+};
+
+async function isDockerImageCreated() {
+  try {
+    const images = await docker.listImages();
+
+    const imageExists = images.some(image => {
+      return image.RepoTags && image.RepoTags.includes(IMAGE_NAME);
+    });
+
+    return imageExists;
+  } catch (error) {
+    console.error("Error checking Docker image:", error);
+    return false;
+  }
+}
+
+async function buildImage() {
+  const dockerfilePath = path.resolve(__dirname, "Dockerfile");
+
+  const command = "docker";
+  const args = [
+    "build",
+    ...BUILD_ARGS,
+    "-t",
+    IMAGE_NAME,
+    "-f",
+    dockerfilePath,
+    cwd,
+  ];
+
+  console.log("Building Docker image...");
+  console.log(`Command: docker ${args.join(" ")}`);
+
+  try {
+    await runCommand(command, args, {
+      onClose: code => {
+        if (code === 0) {
+          return `Docker image built successfully: ${IMAGE_NAME}`;
+        } else {
+          return `Docker build failed with exit code ${code}`;
+        }
+      },
+    });
+  } catch (error) {
+    console.error("Failed to build Docker image:", error.message);
+  }
+}
+
+async function runContainer(options) {
+  const { env = {}, command, after = [] } = options;
+
+  try {
+    container = await docker.createContainer({
+      Image: IMAGE_BASE_NAME,
+      Tty: true,
+      Cmd: ["-c", command, ...after],
+      HostConfig: {
+        Binds: CONTAINER_BOUND_DIRECTORIES.map(
+          directory => `${cwd}/${directory}:${WORKING_DIRECTORY}/${directory}`,
+        ),
+        ExtraHosts: ["host.docker.internal:host-gateway"],
+      },
+      WorkingDir: WORKING_DIRECTORY,
+      Env: Object.entries(env).map(([key, value]) => `${key}=${value}`),
+    });
```

Comment:
For the final implementation, i'll probably will not use `dockerode` and will just use `runCommand` same as when building an image